### PR TITLE
checknonmanualenrolments: Check for all enrolment types to avoid dupes

### PR DESCRIPTION
**Pull request by andrewhancox**
To avoid duplicating enrolments from other plugins optionally check role assignments.

### CODE REVIEW

File: tests/checknonmanualenrolments_test.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Upgrade code for local_mass_enroll
```
Comment (Peterburnett) [inline tests/checknonmanualenrolments_test.php]: This description isn't quite right. Given that there is both a class and file comment, one can also be removed to cut the boilerplate a bit.

File: tests/checknonmanualenrolments_test.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        // 1. All users should have exactly 1 enrolment into the course (i.e. no duplicates)
+        $courseenrol = current(enrol_get_instances($course->id, true));
+
+        for ($i = 1; $i < 4; $i++) {
+            $this->assertCount(1, $DB->get_records('user_enrolments', ['userid' => $users[$i]->id, 'enrolid' => $courseenrol->id]));
+        }
+
+        $this->assertCount($checknonmanualenrolments ? 1 : 2, $DB->get_records('user_enrolments', ['userid' => $alreadyenrolleduser->id]));
+    }
+}
```
Comment (Peterburnett) [inline tests/checknonmanualenrolments_test.php]: Please add a newline to the end of the file

**Comment by Peterburnett** [comment]
Hi Andrew, this looks great, Just a couple of cosmetics. Feel free to ping me again once done and we can get it merged

**Comment by andrewhancox** [comment]
Again, all fair points and done.

**Pull request merged**